contracts/StakeManager.sol

```diff
@@ -185,7 +183,7 @@ contract StakeManager is Ownable {
      * @dev Reverts when amount staked results in less than 1 MP per epoch.
      */
     function stake(uint256 _amount, uint256 _secondsToLock) external onlyVault noPendingMigration {
-        finalizeEpoch();
+        finalizeEpoch(newEpoch());
```

Comment:
This introduces a new problem.
Users don't have control over `newEpoch()`'s size. So if `newEpoch()` is big enough, then the protocol will be break. 

We're now going back full circle to the `limitEpoch` parameter that will have to be provided to every function that calls `finalizeEpoch`.

And even if we did that, there's no guarantee that the given `limitEpoch` will be enough to catch up to where the latest `currentEpoch` **would've been** if epochs were finalized every `EPOCH_SIZE`.

Which then results in the same issue as what I've mentioned in https://github.com/logos-co/staking/pull/127/files#r1770876391

Reply:
To elaborate:

1. Say we're at `currentEpoch = 1`, but from a time stream perspective, we should be at epoch 120
2. User calls `stake(limitEpoch = 20` (imagine `limitEpoch` to be some number that ensures we don't run into block gas limit, but is also `< epoch 120`
3. Protocol finalizes until `epoch 21` and performs stake

Same situation has pointed out in https://github.com/logos-co/staking/pull/127/files#r1770876391

Reply:
I see, that's why there are still functions with _limitEpoch, such as executeEpoch(uint _limitEpoch) and executeAccount(address account, uint _limitEpoch)

Reply:
So if there is a problem with the block gas limit, any address can call executeEpoch(_limitEpoch) to bump the epoch, in case the problem is within processing epoch. And if the problem is with bumping the account userEpoch, than they can call executeAccount(address account, uint _limitEpoch). This _limitEpoch supports smaller chunks... I can add some tests to simulate both cases.

Reply:
> any address can call executeEpoch(_limitEpoch) to bump the epoch, in case the problem is within processing epoch.

That is correct, but if you it still allows you to `stake(limitEpoch)` where `limitEpoch` is `< currentEpoch` then you run into the exact same issue as https://github.com/logos-co/staking/pull/127/files#r1770876391

So if we want to prevent the original issue, we'd have to prevent users from allowing to stake, if their provided `limitEpoch` doesn't each up to `currentEpoch`, forcing them to `processAccount()` first until their epoch is enough that the next attempt to `stake()` works out.

Reply:
That's already the case

Reply:
BTW, the case you mentioned would be possible without these changes. 

Before this changes, the finalizeEpoch would only finalize 1 epoch, even if more time passed. But the problem you describe would not happen without #127 because it wouldnt matter, as finalizeEpoch would always take current block timestamp.   

So having these 2 changes are essential for estimation work properly, and for all the epochs.


I can research other ways on solving this issue, such as taking in consideration for estimation that startTime of epochs can be different than multiples of EPOCH_SIZE.

Reply:
> So having these 2 changes are essential for estimation work properly, and for all the epochs.

Yes. So if we want predictable estimation it looks like we definitely need fixed sized epochs.
If we have fixed sized epochs, then `finalizeEpoch` has to finalize with `EPOCH_SIZE` time segments (like your changes do). Then it automatically aligns with `processAccount()'s` logic.

Now with this ^ we still need to ensure, however, that `stake()` does **not** cause an account to end up in an epoch that is `< currentEpoch`. 

So the way I see it, we need to add a `account.epoch + limitEpochAmount == currentEpoch-1` check. Otherwise we still end up with cases where stake is added "in the past".